### PR TITLE
feat: implement pure http server on listen method

### PR DESCRIPTION
**Pull request by fletcherist**
Design questions on implementation:

- is it okay to return server instance in `listen()` method?

example:
```javascript
const server = alice.listen(8080, '/', () => console.log('listening on 8080'))
```

this enables to use `server.stopListening` directly from developer code

- is `listen()`'s argument's order okay? I think `port` should be the first, `webhookUrl` as the second & optional `callback` is the third.

```javascript
// 1.5.2
alice.listen('/', 8080, () => null)
// 2.0
alice.listen(8080, '/', () => null)
```

I did it because most of the developers don't use `webhookUrl` (they're actually using default `'/'`)

### CODE REVIEW

File: src/alice.ts [hunk 1]
```diff
@@ -81,6 +83,46 @@ export class Alice implements IAlice {
     };
   }
 
+  public listen(
```
Comment (Yavanosta) [inline src/alice.ts]: This function is very long and complex, and gonna grow up if we add https support here. It will be great to split it to several methods. Or create separate class something like AliceHttpServer, move all this logic into it, and just call start\stop methods in alice.ts.

File: src/alice.ts [hunk 2]
```diff
@@ -81,6 +83,46 @@ export class Alice implements IAlice {
...
+              const requestData = Buffer.from(body).toString();
+              if (request.method === 'POST' && request.url === webhookUrl) {
+                try {
+                  const requestBody = JSON.parse(requestData);
+                  const responseBody = await this.handleRequest(requestBody);
+                  response.statusCode = 200;
+                  response.setHeader('Content-Type', 'application/json');
+                  response.end(JSON.stringify(responseBody));
+                } catch (error) {
+                  throw new Error(error);
```
[the remark this answers is not in the document]
Comment (Yavanosta) [reply src/alice.ts]: Besides Error does not have constructor which accept another error, it's accept string message only, so you will loose stack information.
https://developer.mozilla.org/ru/docs/Web/JavaScript/Reference/Global_Objects/Error

File: src/alice.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (fletcherist) [reply src/alice.ts]: thanks! Missed it 👍

File: src/alice.ts [hunk 3]
```diff
@@ -35,16 +38,15 @@ export class Alice implements IAlice {
...
       return null;
     }
 
     let index = middlewares.length - 1;
     const next = async (
       context: IContext,
     ): Promise<IMiddlewareResult | null> => {
-      const middleware = middlewares[index];
-      index--;
+      const middleware = middlewares[index--];
```
[the remark this answers is not in the document]
Comment (fletcherist) [reply src/alice.ts]: Okay, I consider `[index--]` not to be very obvious code. Better use yours then.

File: src/alice.ts [hunk 4]
```diff
@@ -81,6 +83,46 @@ export class Alice implements IAlice {
...
+      .createServer(
+        async (request: http.ServerRequest, response: http.ServerResponse) => {
+          const body: (string | Buffer)[] = [];
+          request
+            .on('data', chunk => {
+              body.push(chunk);
+            })
+            .on('end', async () => {
+              const requestData = Buffer.from(body).toString();
+              if (request.method === 'POST' && request.url === webhookUrl) {
```
[the remark this answers is not in the document]
Comment (fletcherist) [reply src/alice.ts]: Thanks! That's much better

File: src/alice.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (fletcherist) [reply src/alice.ts]: Wow, didn't know about that! We haven't got any idea, why our
```javascript
alice.any(ctx => { qwerty++++++++++- })
```

has stacktrace starts with `../node_modules/yandex-dialogs-sdk/...`,  not in developer's app
now got it. Great

File: src/alice.ts [hunk 5]
```diff
@@ -81,6 +83,46 @@ export class Alice implements IAlice {
     };
   }
 
+  public listen(
+    port: number = 80,
+    webhookUrl: string = '/',
+    callback?: () => void,
```
[the remark this answers is not in the document]
Comment (fletcherist) [reply src/alice.ts]: actually it's quite useless. I use it to just say `console.log('listening on', PORT)`

[the remark this answers is not in the document]
Comment (fletcherist) [reply src/alice.ts]: I consider this to be removed

File: src/alice.ts (hunk 1 above)
Comment (fletcherist) [reply src/alice.ts]: I think https support is not in out interest yet.
Developers should better use `nginx` configuration or CloudFlare for `https`.

But it's rather nice idea to kick out server from main package to make it zen, as it was before.

File: src/alice.ts (hunk 5 above)
[the remark this answers is not in the document]
Comment (fletcherist) [reply src/alice.ts]: i suggest it to be replaced to `options` like that
```javascript
alice.listen(8080, '/', {
 https: true,
 cert: '',
 pem: '',
})
```

File: src/alice.ts [hunk 6]
```diff
@@ -81,6 +83,46 @@ export class Alice implements IAlice {
...
+  ) {
+    this._server = http
+      .createServer(
+        async (request: http.ServerRequest, response: http.ServerResponse) => {
```
[the remark this answers is not in the document]
Comment (fletcherist) [reply src/alice.ts]: Prettier formats as it wants. Can do nothing with this :(

**Comment by fletcherist** [comment]
I moved webhook server to `server/webhookServer` with methods `start(), stop()`. Start is executed by default.
```javascript
const server = alice.listen(8080, '/')
setTimeout(() => server.stop(), 0)
server.start() // Error: server has already been started
```
When they'll implement async protocol (polling/ws/whatsoever), we're ready for this by design.
All that we need is to add `./server/polling.ts`

File: src/alice.ts (hunk 1 above)
Comment (fletcherist) [reply src/alice.ts]: moved to a signle file

File: package.json [hunk 7]
```diff
@@ -49,21 +49,22 @@
   },
   "homepage": "https://github.com/fletcherist/yandex-dialogs-sdk#readme",
   "dependencies": {
-    "@types/node-fetch": "^2.1.2",
     "node-fetch": "^2.1.2"
   },
   "devDependencies": {
-    "@types/express": "^4.16.0",
+    "@types/debug": "^0.0.30",
+    "@types/node-fetch": "^2.1.2",
```
Comment (Yavanosta) [inline package.json]: Are you sure that node fetch types have to be moved here? I beliave that they should be located at same section as node-fetch

File: package.json [hunk 8]
```diff
@@ -49,21 +49,22 @@
...
     "conventional-changelog-cli": "^2.0.0",
     "conventional-github-releaser": "^3.1.0",
     "husky": "^0.14.3",
+    "debug": "^3.1.0",
     "jest": "^23.2.0",
     "prettier": "^1.13.7",
     "ts-jest": "^23.0.0",
     "tslint": "^5.10.0",
-    "typescript": "^2.9.2",
+    "typescript": "^3.0.1",
```
Comment (Yavanosta) [inline package.json]: Please move typescript update to separate PR. At least to separate commit. Commit message does not reflect changes made in this commit.

File: src/alice.ts [hunk 9]
```diff
@@ -1,25 +1,33 @@
...
 import { Middleware, IMiddlewareResult } from './middleware/middleware';
 import { IApiRequest } from './api/request';
 import { IContext } from './context';
 import { IApiResponse } from './api/response';
-import { ALICE_PROTOCOL_VERSION } from './constants';
-import { Reply } from './reply/reply';
+import { ALICE_PROTOCOL_VERSION, LIBRARY_NAME } from './constants';
 
-export interface IAliceConfig extends IImagesApiConfig {}
+const debug = _debug(LIBRARY_NAME);
```
Comment (Yavanosta) [inline src/alice.ts]: Not used variable.

File: src/alice.ts [hunk 10]
```diff
@@ -1,25 +1,33 @@
+import * as http from 'http';
+import _debug from 'debug';
```
Comment (Yavanosta) [inline src/alice.ts]: Not used.

File: src/alice.ts [hunk 11]
```diff
@@ -1,25 +1,33 @@
+import * as http from 'http';
```
Comment (Yavanosta) [inline src/alice.ts]: not used

File: src/alice.ts [hunk 12]
```diff
@@ -1,25 +1,33 @@
...
+export interface IAliceConfig extends IImagesApiConfig {
+  oAuthToken?: string;
+  skillId?: string;
+}
 
 export interface IAlice {
   readonly imagesApi: IImagesApi;
   handleRequest(data: IApiRequest): Promise<IApiResponse>;
   use(middleware: Middleware): void;
+  listen(port: number, webhookUrl: string, options: object): IWebhookServer;
```
Comment (Yavanosta) [inline src/alice.ts]: I believe that you have to create interface for options.

File: src/alice.ts [hunk 13]
```diff
@@ -81,6 +89,21 @@ export class Alice implements IAlice {
...
+  public listen(
+    port: number = 80,
+    webhookUrl: string = '/',
+    options: object = {},
+  ): WebhookServer {
+    const server = new WebhookServer({
+      port: port,
+      webhookUrl: webhookUrl,
+      options: options,
+      handleRequest: this.handleRequest,
```
Comment (Yavanosta) [inline src/alice.ts]: Don't you loose `this` context here? Maybe `handleRequest: request => this.handleRequest(request)` or something like that?

File: src/index.ts [hunk 14]
```diff
@@ -9,6 +9,16 @@ export {
   CommandMatcher,
 } from './command/command';
 
+import { Alice } from './alice';
```
Comment (Yavanosta) [inline src/index.ts]: I believe we need separate commit or PR for this.

File: src/server/webhookServer.ts [hunk 15]
```diff
@@ -0,0 +1,74 @@
...
+
+const debug = _debug(LIBRARY_NAME);
+
+export interface IWebhookServerConfig {
+  port: number;
+  webhookUrl: string;
+  options: object;
+  handleRequest: IAlice['handleRequest'];
+}
+export interface IWebhookServer {
```
Comment (Yavanosta) [inline src/server/webhookServer.ts]: Empty line

File: src/server/webhookServer.ts [hunk 16]
```diff
@@ -0,0 +1,74 @@
...
+import { LIBRARY_NAME } from '../constants';
+import { IAlice } from '../alice';
+
+const debug = _debug(LIBRARY_NAME);
+
+export interface IWebhookServerConfig {
+  port: number;
+  webhookUrl: string;
+  options: object;
+  handleRequest: IAlice['handleRequest'];
```
Comment (Yavanosta) [inline src/server/webhookServer.ts]: It's not a good idea to link IAlice and IWebhookServerConfig intefaces.

Better something like (request: IApiRequest): IApiResponse

File: src/server/webhookServer.ts [hunk 17]
```diff
@@ -0,0 +1,74 @@
...
+}
+export interface IWebhookServer {
+  start(): void;
+  stop(): void;
+}
+
+export class WebhookServer {
+  private server: http.Server;
+  private port: number;
```
Comment (Yavanosta) [inline src/server/webhookServer.ts]: Please add underscore before private fields. As soon as this lib will be used not only in type script i think it's good idea to have private marker not only as field modificator, but as name prefix.

File: src/server/webhookServer.ts [hunk 18]
```diff
@@ -0,0 +1,74 @@
...
+      throw new Error(`Server is already started`);
+    }
+    this.server.listen(this.port, () => {
+      debug(`server is listening on ${this.port}, '${this.webhookUrl}'`);
+      this.isServerStarted = true;
+    });
+  }
+
+  public stop(): void {
+    if (this.isServerStarted) {
```
Comment (Yavanosta) [inline src/server/webhookServer.ts]: It's up to you but it's good idea to invert `if` in such cases.

```
if (!this.isServerStarted) {
  return;
}
debug(`stopping webhook server`);
this.server.close();
this.isServerStarted = false;

File: src/server/webhookServer.ts [hunk 19]
```diff
@@ -0,0 +1,74 @@
...
+  private port: number;
+  private webhookUrl: string;
+  private isServerStarted: boolean;
+  private handleRequest: IAlice['handleRequest'];
+
+  constructor(config: IWebhookServerConfig) {
+    this.port = config.port;
+    this.webhookUrl = config.webhookUrl;
+    this.handleRequest = config.handleRequest;
+    this.isServerStarted = false;
```
Comment (Yavanosta) [inline src/server/webhookServer.ts]: You dont need word server here. `_isStarted` is enough i think. But it's up to you.

File: src/server/webhookServer.ts [hunk 20]
```diff
@@ -0,0 +1,74 @@
...
+            }
+            const requestBody = JSON.parse(requestData);
+            const responseBody = await this.handleRequest(requestBody);
+            response.statusCode = 200;
+            response.setHeader('Content-Type', 'application/json');
+            response.end(JSON.stringify(responseBody));
+          });
+      },
+    );
+    return this;
```
Comment (Yavanosta) [inline src/server/webhookServer.ts]: You don't need to return something from constructor.

File: src/server/webhookServer.ts [hunk 21]
```diff
@@ -0,0 +1,74 @@
...
+  private handleRequest: IAlice['handleRequest'];
+
+  constructor(config: IWebhookServerConfig) {
+    this.port = config.port;
+    this.webhookUrl = config.webhookUrl;
+    this.handleRequest = config.handleRequest;
+    this.isServerStarted = false;
+
+    debug(`create server on: ${this.port}, ${this.webhookUrl}`);
+    this.server = http.createServer(
```
Comment (Yavanosta) [inline src/server/webhookServer.ts]: I think we can split this function one more time (something like this):
```
constructor {
    this.server = http.createServer(
        (request, response) => this._handleRequest(request, response));
}

private async _handleRequest(request, response) {
   const requestData = await AliceHttpServer._readRequest(request);
   const responseData = await this.handleRequestData_(requestData);
   await AliceHttpServer._sendResponse(response, responseData);
}

private static async _sendResponse(response, responseBody: IApiReposne): Promise {
   response.statusCode = 200;
   response.setHeader('Content-Type', 'application/json');
   response.end(JSON.stringify(responseBody));
}

private static async _readRequest(request): Promise<IApiRequest> {
    return new Promise((resolve, reject) => {
      const body: Array<Buffer> = [];
      request
            .on('data', chunk => body.push(chunk))
            .on('end', () => {
                if (request.method !== 'POST' || request.url !== this.webhookUrl) {
                   response.statusCode = 400;
                   response.end();
                   reject();
                }
                resolve(JSON.parse<IApiRequest>(requestData));
            })
            .on('error', reject); // I don't know if there is something like this.
    });
}
```

File: package.json (hunk 7 above)
Comment (fletcherist) [reply package.json]: thanks

File: src/server/webhookServer.ts (hunk 19 above)
Comment (fletcherist) [reply src/server/webhookServer.ts]: hmm, okay, `_isStarted` seems to be better one

File: src/server/webhookServer.ts (hunk 20 above)
Comment (fletcherist) [reply src/server/webhookServer.ts]: thx! didn't catch it

File: src/server/webhookServer.ts (hunk 18 above)
Comment (fletcherist) [reply src/server/webhookServer.ts]: 👍  thnx!

File: src/server/webhookServer.ts (hunk 15 above)
Comment (fletcherist) [reply src/server/webhookServer.ts]: thank you

File: src/server/webhookServer.ts (hunk 17 above)
Comment (fletcherist) [reply src/server/webhookServer.ts]: done. Actually I'm not true fan of underscores in private methods, bc this `webhookServer` won't be used outside the library, but anyways, okay, maybe maybe

File: src/server/webhookServer.ts (hunk 16 above)
Comment (fletcherist) [reply src/server/webhookServer.ts]: done

File: src/index.ts (hunk 14 above)
Comment (fletcherist) [reply src/index.ts]: excluded in a separate PR

File: src/server/webhookServer.ts (hunk 21 above)
Comment (fletcherist) [reply src/server/webhookServer.ts]: Splitted into separate functions and refactored

File: package.json [hunk 22]
```diff
@@ -49,16 +49,17 @@
   },
   "homepage": "https://github.com/fletcherist/yandex-dialogs-sdk#readme",
   "dependencies": {
-    "@types/node-fetch": "^2.1.2",
-    "node-fetch": "^2.1.2"
+    "node-fetch": "^2.1.2",
+    "@types/node-fetch": "^2.1.2"
```
Comment (Yavanosta) [inline package.json]: Not in alphabetical order

File: package.json [hunk 23]
```diff
@@ -49,16 +49,17 @@
...
   },
   "devDependencies": {
-    "@types/express": "^4.16.0",
+    "@types/debug": "^0.0.30",
     "@types/jest": "^23.1.3",
     "@types/node": "^10.5.1",
     "conventional-changelog-cli": "^2.0.0",
     "conventional-github-releaser": "^3.1.0",
     "husky": "^0.14.3",
+    "debug": "^3.1.0",
```
Comment (Yavanosta) [inline package.json]: Not in alphabetical

File: src/alice.ts [hunk 24]
```diff
@@ -1,25 +1,29 @@
...
 import { Middleware, IMiddlewareResult } from './middleware/middleware';
 import { IApiRequest } from './api/request';
 import { IContext } from './context';
 import { IApiResponse } from './api/response';
 import { ALICE_PROTOCOL_VERSION } from './constants';
-import { Reply } from './reply/reply';
 
-export interface IAliceConfig extends IImagesApiConfig {}
+export interface IAliceConfig extends IImagesApiConfig {
+  oAuthToken?: string;
```
Comment (Yavanosta) [inline src/alice.ts]: Но ведь это уже есть в IImagesApiConfig зачем дублировать?

File: src/server/webhookServer.ts [hunk 25]
```diff
@@ -0,0 +1,95 @@
+import * as http from 'http';
+import debug from '../debug';
+import { IApiRequest } from '../api/request';
+import { IApiResponse } from '../api/response';
+
+type handleAliceRequestType = (request: IApiRequest) => Promise<IApiResponse>;
```
Comment (Yavanosta) [inline src/server/webhookServer.ts]: Start type names with capital letter

File: src/server/webhookServer.ts [hunk 26]
```diff
@@ -0,0 +1,95 @@
...
+      const body: Array<string | Buffer> = [];
+      request
+        .on('data', chunk => {
+          body.push(chunk);
+        })
+        .on('end', async () => {
+          const requestData = Buffer.from(body).toString();
+          resolve(JSON.parse(requestData));
+        })
+        .on('error', reject);
```
Comment (Yavanosta) [inline src/server/webhookServer.ts]: There is no 'error' event on request. :-(
Sorry. You can leave it as is i'll fix it later.

File: tslint.json [hunk 27]
```diff
@@ -14,6 +14,7 @@
 		"arrow-parens": false,
 		"object-literal-shorthand": false,
 		"only-arrow-functions": false,
+		"member-ordering": false
```
Comment (Yavanosta) [inline tslint.json]: Why you disabled this?

File: src/server/webhookServer.ts (hunk 25 above)
Comment (fletcherist) [reply src/server/webhookServer.ts]: ty, missed that stuff
